docs/honk.3
.\"
.\" Copyright (c) 2019 Ted Unangst
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt HONK 3
.Os
.Sh NAME
.Nm honk
.Nd API access
.Sh DESCRIPTION
In addition to the standard web interface, some functionality is
available via the
.Nm
HTTP API.
.Pp
With the exception of login, all requests should contain
the following form values.
.Bl -tag -width action
.It Fa action
The desired action.
See below.
.It Fa token
An authorization token.
Alternatively, may be passed in the
.Dq Authorization
HTTP header.
.El
.Pp
The API URL for all actions other than login and logout is
.Pa /api .
.Ss login
Send a POST request to
.Pa /dologin
with the following form values.
.Bl -tag -width username
.It Fa username
User name.
.It Fa password
Pass phrase.
.It Fa gettoken
Must be
.Dq 1 .
.El
.Pp
This will return a token to be used for future requests.
The token is valid for one year.
.Ss logout
Send a request to
.Pa /logout
with the
.Fa token
to be expired.
.Ss honk
The
.Fa action
value should be
.Dq honk .
Content type should be multipart/form-data if an attachment is included.
The following values are recognized:
.Bl -tag -width placename
.It Fa noise
The contents of the honk.
.It Fa format
The format of noise.
Defaults to markdown.
May also be html.
.It Fa donk
A file to attach.
.It Fa donkdesc
A description for the attached file.
.It Fa donkxid
The XID of a previously uploaded attachment.
.It Fa placename
The name of an associated location.
.It Fa placeurl
The url of an associated location.
.It Fa placelat
The latitude of an associated location.
.It Fa placelong
The longitude of an associated location.
.It Fa timestart
The start time of an event.
.It Fa rid
The ActivityPub ID that this honk is in reply to.
.El
.Pp
Upon success, the honk action will return the URL for the created honk.
.Ss donk
Upload just an attachment using
.Fa donk
and
.Fa donkdesc .
Content type must be multipart/form-data.
Will return the XID.
.Ss gethonks
The
.Dq gethonks
.Fa action
can be used to query for honks.
The following parameters are used.
.Bl -tag -width placename
.It Fa page
Should be one of
.Dq home
or
.Dq atme .
.It Fa after
Only return honks after the specified ID.
.It Fa wait
If there are no results, wait this many seconds for something to appear.
.El
.Pp
The result will be returned as json.
.Ss zonkit
The
.Dq zonkit
action began life as a delete function, but has since evolved some other
powers as specified by the
.Fa wherefore
parameter.
The target of the action is specified by the
.Fa what
parameter and is generally the XID of a honk.
.Pp
Wherefore must be one of the following.
.Bl -tag -width zonvoy
.It bonk
Share honk with others.
.It unbonk
Undo share.
.It save
Mark honk as saved.
.It unsave
Unmark honk as saved.
.It react
Post an emoji reaction.
A custom reaction may be specified with
.Fa reaction .
.It ack
Mark honk as read.
.It deack
Unmark honk as read.
.It zonk
Delete this honk.
.It zonvoy
Mute this thread.
What should identify a convoy.
.Ss gethonkers
Returns a list of current honkers in json format.
.Ss savehonker
Save a new honker, or update an existing one.
The following fields are used.
.Bl -tag -width honkerid
.It Fa honkerid
The numeric ID of an existing honker to update.
.It Fa name
The preferred short name.
.It Fa url
The ActivityPub actor url.
.It Fa combos
Space separated list of combos.
.It Fa notes
Some notes.
.El
.Pp
The honker numeric ID will be returned for success.
To delete, unsub, or sub, include a form value with name and value equal.
As in, a form value named delete with the value delete, or unsub=unsub, etc.
.El
.Ss sendactivity
Send anything.
No limits, no error checking.
.Bl -tag -width public
.It Fa rcpt
An actor to deliver the message to to.
May be specified more than once.
An inbox may be specified directly by prefixing with %.
.It Fa msg
The message.
It should be a valid json activity, but yolo.
.It Fa public
Set to 1 to use shared inboxes for delivery.
.El
.Sh EXAMPLES
Refer to the sample code in the
.Pa toys
directory.
.Sh SEE ALSO
.Xr vim 3


docs/changelog.txt
changelog

=== next

+ All inclusive danger zone spoiler alerts.

+ Filter to match anything with summary/warning.

+ Start collecting quties.

+ Fix http signatures for GET requests.

+ Fix adjacent mentions.

+ Fix argv for chpass.

+ Avoid self mention in reply all.

+ Fix markdown links with parens.

+ Add a default icon.png.

+ Try to fix hoot again because Twitter did a Twitter.

=== 0.9.8 Tentative Tentacle

+ Switch database to WAL mode.

- go version 1.16 required.

+ Specify banner: image in profile.

+ Update activity compatibility with mastodon.

- Signed fetch.

+ Better unicode hashtags.

+ Some more configuration options.

+ Some UI improvements to web interface.

+ Add atme class to mentions

+ Improvements to the mastodon importer.

+ More hydration capable pages.

+ Support for local.js.

+ Better error messages for timeouts.

+ Some improved html and markdown.

=== 0.9.7 Witless Weekender

+++ Word guessing game. Wonk wonk!

+ Flexible logging, to file, syslog, null, etc.

+ Low key unread counters.

+ Images in the hooter.

+ More flexible hashtag characters.

+ Fix the memetizer to work in more environments.

+ Printing is prettier than ever before.

=== 0.9.6 Virile Vigorous and Potent

+ A bug, a fix, a bug fix, a fix bug.

+ Fix Update processing.

+ Better cookie rotation with weekly refresh.

+ A new follow button in a surprise location.

+ Fix mastodon import.

+ Filters work better with hashtags.

+ Fix hoot to work with Twitter's latest crap.

=== 0.9.5 Emergency Ejection

+ Fix honk init user creation.

=== 0.9.4 Collegiate Colloquialism

+ Add validation to some more user inputs to prevent mistakes.

+ Easier to use ping command.

=== 0.9.3 Notacanthous Nutshell

++ backup command.

+ Relax requirement for multipart/form-data posts in API.

+ Dedupe blob file data.

+ Better support for rich text bios.

+ Follow and unfollow should work a little better.

+ Option to mention all in replies.

+ Reduce interference between various text substitution rules.

+ Fix crash in search with extra space.

+ Fix pubkey issue with domain only keys.

- Custom lingo for those who don't like honking.

=== 0.9.2 Malleable Maltote

+ Fix compilation on mac.

=== 0.9.1 Late Stage Lusciousness

++ Boing boom tschak chonky chatter. Chat messages with Pleroma.

+ Custom rgb flag: emoji.

+ Slightly better ActivityPub compat

+ ## headings for markdown

+ Workaround js only twitter for hoot: feature.

+ Quote unquote reliability improvements.

+ Much better omit images handling.

+ Fix update activity.

+ A few API refinements and additions.

=== 0.9.0 Monitor vs Merrimack

--- Add Reactions.

+++ Rename react to badonk.

+ Quick fix to hide all images.

+ Allow resending follow requests.

+ Improved search query parsing.

+ Tables

+ Reduce retries talking to dumb servers.

+ Maybe possible to use @user@example.com wihtout subdomain.

=== 0.8.6 Sartorial Headpiece

++ Import command now supports the elephant in the room.

+ Minimal support for Move activity.

+ deluser command.

+ Configurable avatar colors.

+ Optional pleroma color scheme for the home sick...

+ Rebalance colors slightly. Looks a little fresher now?

+ Add unplug command for servers that have dropped off the net.

+ Add notes field to honkers to document their downfall.

+ Add notes field to filters for record keeping.

+ Negated search -terms.

+ A raw sendactivity API action for the bold.

+ More flexible meme names.

=== 0.8.5 Turnkey Blaster

+ Codenames in changelog.

+ Fix some bugs that may have interfered with federation.

+ Add some re: re: re: to replies.

+ Set an avatar. If you must.

+ Try a little harder to recover from httpsig failures.

+ Add cite tag for block quote attributions.

+ @media print styles.

+ Disable overscroll (pull down) refresh.

+ Can never seem to version the changelog correctly.

=== 0.8.4

+ Fix bug preventing import of keys

+ Option to switch map links to Apple.

=== 0.8.3

- mistag.

=== 0.8.2 Game Warden

++ Import command to preserve those embarssassing old posts from Twitter.

++ Add a limited /api for the robotrons.

+ Resource usage stats on about page.

+ Unveil and pledge restrictions on OpenBSD.

+ Lists supported in markdown.

+ Rewrite admin console to avoid large dependencies.

+ "Bug" fixes.

=== 0.8.1

++ Make it easier to upgrade by decoupling data dir from ".".

+ Timestamps displayed in server time with TZ.

+ version command to print current version.

+ Amend changelog for 0.8.0 to include omitted elements:
	Syntax highlighting for code blocks.
	Something resembling an actual manual.

=== 0.8.0 Ordinary Octology

+++ Add Honk Filtering and Censorship System (HFCS).

+++ Editing honks (Update activity).

++ Subscribe to hashtags.

++ Search. I hate it already.

++ Hashtags that work?

++ Dynamic refresh and page switching without reloads.

++ Reply control. Ack replies to show them on the site.

+ Allow PDF attachments. For serious business only.

+ "untag me" button to mute part of a thread.

+ Inline images in posts. Send and receive.

+ Somewhat functional admin console (TTY).

+ More JS free fallbacks for some basic functions.

+ Add chpass command.

+ Improved honker management.

+ Better markdown output.

+ Times for events.

+ Split media database into separate blob.db.

+ Location checkin. Welcome to the... danger zone!

+ Quick mention @alias.

+ Image descriptions.

+ Unbonking.

+ More robust retries for fetching objects.

+ Don't decode excessively large images and run out of memory.

+ Syntax highlighting for code blocks.

+ Something resembling an actual manual.

- Sometimes the cached state of the @me feed becomes unsynced.
	Acked status may display incorrectly.

=== 0.7.7 More 7 Than Ever

+ Add another retry to workaround pixelfed's general unreliability.

+ Attached images are not lost when previewing.

- Remove sensitivity to spicy peppers.

+ Keep reply to setting during preview.

+ Increase max thread retrieval depth to 10.


honk.go
//
// Copyright (c) 2019 Ted Unangst <tedu@tedunangst.com>
//
// Permission to use, copy, modify, and distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

package main

import (
	"flag"
	"fmt"
	"html/template"
	golog "log"
	"log/syslog"
	notrand "math/rand"
	"os"
	"strconv"
	"strings"
	"time"

	"humungus.tedunangst.com/r/webs/httpsig"
	"humungus.tedunangst.com/r/webs/log"
	"humungus.tedunangst.com/r/webs/mz"
)

var softwareVersion = "fisherman-mod"

func init() {
	notrand.Seed(time.Now().Unix())
}

type WhatAbout struct {
	ID      int64
	Name    string
	Display string
	About   string
	HTAbout template.HTML
	Onts    []string
	Key     string
	URL     string
	Options UserOptions
	SecKey  httpsig.PrivateKey
}

type UserOptions struct {
	SkinnyCSS    bool   `json:",omitempty"`
	OmitImages   bool   `json:",omitempty"`
	MentionAll   bool   `json:",omitempty"`
	InlineQuotes bool   `json:",omitempty"`
	Avatar       string `json:",omitempty"`
	Banner       string `json:",omitempty"`
	MapLink      string `json:",omitempty"`
	Reaction     string `json:",omitempty"`
	MeCount      int64
	ChatCount    int64
}

type KeyInfo struct {
	keyname string
	seckey  httpsig.PrivateKey
}

const serverUID int64 = -2
const readyLuserOne int64 = 1

type Honk struct {
	ID       int64
	UserID   int64
	Username string
	What     string
	Honker   string
	Handle   string
	Handles  string
	Oonker   string
	Oondle   string
	XID      string
	RID      string
	Date     time.Time
	URL      string
	Noise    string
	Precis   string
	Format   string
	Convoy   string
	Audience []string
	Public   bool
	Whofore  int64
	Replies  []*Honk
	Flags    int64
	HTPrecis template.HTML
	HTML     template.HTML
	Style    string
	Open     string
	Donks    []*Donk
	Onts     []string
	Place    *Place
	Time     *Time
	Mentions []Mention
	Badonks  []Badonk
}

type Badonk struct {
	Who  string
	What string
}

type Chonk struct {
	ID     int64
	UserID int64
	XID    string
	Who    string
	Target string
	Date   time.Time
	Noise  string
	Format string
	Donks  []*Donk
	Handle string
	HTML   template.HTML
}

type Chatter struct {
	Target string
	Chonks []*Chonk
}

type Mention struct {
	Who   string
	Where string
}

func (mention *Mention) IsPresent(noise string) bool {
	nick := strings.TrimLeft(mention.Who, "@")
	idx := strings.IndexByte(nick, '@')
	if idx != -1 {
		nick = nick[:idx]
	}
	return strings.Contains(noise, ">@"+nick) || strings.Contains(noise, "@<span>"+nick)
}

type OldRevision struct {
	Precis string
	Noise  string
}

const (
	flagIsAcked    = 1
	flagIsBonked   = 2
	flagIsSaved    = 4
	flagIsUntagged = 8
	flagIsReacted  = 16
	flagIsWonked   = 32
)

func (honk *Honk) IsAcked() bool {
	return honk.Flags&flagIsAcked != 0
}

func (honk *Honk) IsBonked() bool {
	return honk.Flags&flagIsBonked != 0
}

func (honk *Honk) IsSaved() bool {
	return honk.Flags&flagIsSaved != 0
}

func (honk *Honk) IsUntagged() bool {
	return honk.Flags&flagIsUntagged != 0
}

func (honk *Honk) IsReacted() bool {
	return honk.Flags&flagIsReacted != 0
}

type Donk struct {
	FileID   int64
	XID      string
	Name     string
	Desc     string
	URL      string
	Media    string
	Local    bool
	External bool
}

type Place struct {
	Name      string
	Latitude  float64
	Longitude float64
	Url       string
}

type Duration int64

func (d Duration) String() string {
	s := time.Duration(d).String()
	if strings.HasSuffix(s, "m0s") {
		s = s[:len(s)-2]
	}
	if strings.HasSuffix(s, "h0m") {
		s = s[:len(s)-2]
	}
	return s
}

func parseDuration(s string) time.Duration {
	didx := strings.IndexByte(s, 'd')
	if didx != -1 {
		days, _ := strconv.ParseInt(s[:didx], 10, 0)
		dur, _ := time.ParseDuration(s[didx:])
		return dur + 24*time.Hour*time.Duration(days)
	}
	dur, _ := time.ParseDuration(s)
	return dur
}

func msgFromMd(s string, msg *template.HTML) {
	var marker mz.Marker
	md, err := os.ReadFile(dataDir+"/"+s+".md")
	if err != nil {
		getconfig(s+"msg", msg)
	} else {
		*msg = template.HTML(marker.Mark(string(md)))
	}
}

type Time struct {
	StartTime time.Time
	EndTime   time.Time
	Duration  Duration
}

type Honker struct {
	ID     int64
	UserID int64
	Name   string
	XID    string
	Handle string
	Flavor string
	Combos []string
	Meta   HonkerMeta
}

type HonkerMeta struct {
	Notes string
}

type SomeThing struct {
	What  int
	XID   string
	Owner string
	Name  string
}

const (
	SomeNothing int = iota
	SomeActor
	SomeCollection
)

var serverName string
var serverPrefix string
var masqName string
var dataDir = "."
var viewDir = "."
var iconName = "icon.png"
var serverMsg template.HTML
var aboutMsg template.HTML
var loginMsg template.HTML

func ElaborateUnitTests() {
}

func unplugserver(hostname string) {
	db := opendatabase()
	xid := fmt.Sprintf("%%https://%s/%%", hostname)
	db.Exec("delete from honkers where xid like ? and flavor = 'dub'", xid)
	db.Exec("delete from doovers where rcpt like ?", xid)
}

func reexecArgs(cmd string) []string {
	args := []string{"-datadir", dataDir}
	args = append(args, log.Args()...)
	args = append(args, cmd)
	return args
}

var elog, ilog, dlog *golog.Logger

func main() {
	flag.StringVar(&dataDir, "datadir", dataDir, "data directory")
	flag.StringVar(&viewDir, "viewdir", viewDir, "view directory")
	flag.Parse()

	log.Init(log.Options{Progname: "honk", Facility: syslog.LOG_UUCP})
	elog = log.E
	ilog = log.I
	dlog = log.D

	args := flag.Args()
	cmd := "run"
	if len(args) > 0 {
		cmd = args[0]
	}
	switch cmd {
	case "init":
		initdb()
	case "upgrade":
		upgradedb()
	case "version":
		fmt.Println(softwareVersion)
		os.Exit(0)
	}
	db := opendatabase()
	dbversion := 0
	getconfig("dbversion", &dbversion)
	if dbversion != myVersion {
		elog.Fatal("incorrect database version. run upgrade.")
	}

	msgFromMd("server", &serverMsg)
	msgFromMd("about", &aboutMsg);
	msgFromMd("login", &loginMsg)

	getconfig("servername", &serverName)
	getconfig("masqname", &masqName)
	if masqName == "" {
		masqName = serverName
	}
	serverPrefix = fmt.Sprintf("https://%s/", serverName)
	getconfig("usersep", &userSep)
	getconfig("honksep", &honkSep)
	getconfig("devel", &develMode)
	getconfig("fasttimeout", &fastTimeout)
	getconfig("slowtimeout", &slowTimeout)
	getconfig("signgets", &signGets)
	prepareStatements(db)
	switch cmd {
	case "admin":
		adminscreen()
	case "import":
		if len(args) != 4 {
			elog.Fatal("import username mastodon|twitter srcdir")
		}
		importMain(args[1], args[2], args[3])
	case "devel":
		if len(args) != 2 {
			elog.Fatal("need an argument: devel (on|off)")
		}
		switch args[1] {
		case "on":
			setconfig("devel", 1)
		case "off":
			setconfig("devel", 0)
		default:
			elog.Fatal("argument must be on or off")
		}
	case "setconfig":
		if len(args) != 3 {
			elog.Fatal("need an argument: setconfig key val")
		}
		var val interface{}
		var err error
		if val, err = strconv.Atoi(args[2]); err != nil {
			val = args[2]
		}
		setconfig(args[1], val)
	case "adduser":
		adduser()
	case "deluser":
		if len(args) < 2 {
			fmt.Printf("usage: honk deluser username\n")
			return
		}
		deluser(args[1])
	case "chpass":
		if len(args) < 2 {
			fmt.Printf("usage: honk chpass username\n")
			return
		}
		chpass(args[1])
	case "follow":
		if len(args) < 3 {
			fmt.Printf("usage: honk follow username url\n")
			return
		}
		user, err := butwhatabout(args[1])
		if err != nil {
			fmt.Printf("user not found\n")
			return
		}
		var meta HonkerMeta
		mj, _ := jsonify(&meta)
		honkerid, err := savehonker(user, args[2], "", "presub", "", mj)
		if err != nil {
			fmt.Printf("had some trouble with that: %s\n", err)
			return
		}
		followyou(user, honkerid, true)
	case "unfollow":
		if len(args) < 3 {
			fmt.Printf("usage: honk unfollow username url\n")
			return
		}
		user, err := butwhatabout(args[1])
		if err != nil {
			fmt.Printf("user not found\n")
			return
		}
		row := db.QueryRow("select honkerid from honkers where xid = ? and userid = ? and flavor in ('sub')", args[2], user.ID)
		var honkerid int64
		err = row.Scan(&honkerid)
		if err != nil {
			fmt.Printf("sorry couldn't find them\n")
			return
		}
		unfollowyou(user, honkerid, true)
	case "cleanup":
		arg := "30"
		if len(args) > 1 {
			arg = args[1]
		}
		cleanupdb(arg)
	case "unplug":
		if len(args) < 2 {
			fmt.Printf("usage: honk unplug servername\n")
			return
		}
		name := args[1]
		unplugserver(name)
	case "backup":
		if len(args) < 2 {
			fmt.Printf("usage: honk backup dirname\n")
			return
		}
		name := args[1]
		svalbard(name)
	case "ping":
		if len(args) < 3 {
			fmt.Printf("usage: honk ping (from username) (to username or url)\n")
			return
		}
		name := args[1]
		targ := args[2]
		user, err := butwhatabout(name)
		if err != nil {
			elog.Printf("unknown user")
			return
		}
		ping(user, targ)
	case "run":
		serve()
	case "backend":
		backendServer()
	case "test":
		ElaborateUnitTests()
	default:
		elog.Fatal("unknown command")
	}
}
